tests/priority_queue_test.c

```diff
@@ -0,0 +1,114 @@
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/priority_queue.h>
+#include <aws_test_harness.h>
+#include <stdlib.h>
+
+static int int_compare(void *a, void *b) {
+    int *a_val = (int*)a;
+    int *b_val = (int*)b;
+
+    return *a_val > *b_val; // min heap
+}
+
+static int test_priority_queue_preserves_order(struct aws_allocator *alloc, void *ctx) {
+    struct aws_priority_queue queue;
+
+    aws_priority_queue_dynamic_init(&queue, alloc, 10, sizeof(int), int_compare);
+
+    int first = 45, second = 67, third = 80, fourth = 120, fifth = 10000;
+
+    aws_priority_queue_push(&queue, &third);
+    aws_priority_queue_push(&queue, &fourth);
+    aws_priority_queue_push(&queue, &second);
+    aws_priority_queue_push(&queue, &fifth);
+    aws_priority_queue_push(&queue, &first);
+
+    size_t num_elements = aws_priority_queue_size(&queue);
+    ASSERT_INT_EQUALS(5, num_elements, "Priority queue size should have been %d but was %d", 5, num_elements);
+
+    int pop_val, top_val;
+    aws_priority_queue_top(&queue, &top_val);
+    aws_priority_queue_pop(&queue, &pop_val);
+    ASSERT_INT_EQUALS(first, pop_val, "First element returned should have been %d but was %d", first, pop_val);
+    ASSERT_INT_EQUALS(pop_val, top_val, "Popped element should have been the top element. expected %d but was %d",
+            pop_val, top_val);
+
+    aws_priority_queue_top(&queue, &top_val);
+    aws_priority_queue_pop(&queue, &pop_val);
+    ASSERT_INT_EQUALS(second, pop_val, "Second element returned should have been %d but was %d", second, pop_val);
+    ASSERT_INT_EQUALS(pop_val, top_val, "Popped element should have been the top element. expected %d but was %d",
+            pop_val, top_val);
+
+    aws_priority_queue_top(&queue, &top_val);
+    aws_priority_queue_pop(&queue, &pop_val);
+    ASSERT_INT_EQUALS(third, pop_val, "Third element returned should have been %d but was %d", third, pop_val);
+    ASSERT_INT_EQUALS(pop_val, top_val, "Popped element should have been the top element. expected %d but was %d", 
+            pop_val, top_val);
+
+    aws_priority_queue_top(&queue, &top_val);
+    aws_priority_queue_pop(&queue, &pop_val);
+    ASSERT_INT_EQUALS(fourth, pop_val, "Fourth element returned should have been %d but was %d", fourth, pop_val);
+    ASSERT_INT_EQUALS(pop_val, top_val, "Popped element should have been the top element. expected %d but was %d",
+            pop_val, top_val);
+
+    aws_priority_queue_top(&queue, &top_val);
+    aws_priority_queue_pop(&queue, &pop_val);
+    ASSERT_INT_EQUALS(fifth, pop_val, "Fifth element returned should have been %d but was %d", fifth, pop_val);
+
+    ASSERT_INT_EQUALS(AWS_ERROR_PRIORITY_QUEUE_EMPTY, aws_priority_queue_pop(&queue, &pop_val), "Popping from empty queue should result in error");
```

Comment:
ASSERT_ERROR, we shouldn't be returning error codes, we return AWS_OP_ERR, and aws_last_error() exposes the error. the ASSERT_ERROR macro will do this check for you.

Reply:
ok thanks